File: apps/ElixirPlug/lib/elixir_plug_api/ws_orders_handler.ex
defmodule ElixirPlugApi.WsOrdersHandler do
  @behaviour :cowboy_websocket

  def init(req, state) do
    {:cowboy_websocket, req, state, %{idle_timeout: 60_000}}
  end

  def websocket_handle({:text, _}, state) do
    query =
      "SELECT id, customer_id, total_cents, status, created_at FROM orders LIMIT $1 OFFSET $2"

    case Postgrex.query(:db, query, [100, 1000]) do
      {:ok, result} ->
        orders =
          Enum.map(result.rows, fn [id, cid, tc, st, ca] ->
            %{
              id: id,
              customer_id: cid,
              total_cents: tc,
              status: st,
              created_at: ca
            }
          end)

        json = Jason.encode!(orders)
        {:reply, {:text, json}, state}

      {:error, _} ->
        {:ok, state}
    end
  end

  def websocket_handle(_, state) do
    {:ok, state}
  end

  def websocket_info(_, state) do
    {:ok, state}
  end
end

File: apps/ErlangCowboy/src/ws_orders_handler.erl
-module(ws_orders_handler).
-behaviour(cowboy_websocket).

-export([init/2, websocket_handle/2, websocket_info/2]).

init(Req, State) ->
    {cowboy_websocket, Req, State}.

websocket_handle({text, _}, State) ->
    Sql = "SELECT id, customer_id, total_cents, status, created_at FROM orders LIMIT $1 OFFSET $2",
    case db_pool:query(Sql, [100, 1000]) of
        {ok, _Columns, Rows} ->
            Orders = lists:map(fun orders_handler:row_to_order/1, Rows),
            Json = jsx:encode(Orders),
            {reply, {text, Json}, State};
        {error, _} ->
            {ok, State}
    end;
websocket_handle(_, State) ->
    {ok, State}.

websocket_info(_, State) ->
    {ok, State}.
